metron-platform/metron-common/src/test/java/org/apache/metron/common/dsl/functions/StringFunctionsTest.java

```diff
@@ -32,124 +33,143 @@
-        }
-        Assert.assertTrue(thrown);
-        thrown = false;
-
-        // NON INTEGER LENGTH
-        try {
-            run("FILL_RIGHT('123','X', 'z' )", new HashedMap());
-        }catch(ParseException pe){
-            thrown = true;
-            Assert.assertTrue(pe.getMessage().contains("not a valid Integer"));
-        }
-        Assert.assertTrue(thrown);
-        thrown = false;
-
-        // EMPTY STRING PAD
-        try {
-            Object returnValue = run("FILL_RIGHT('123','', 10 )", new HashedMap());
-        }catch(ParseException pe) {
-            thrown = true;
-            Assert.assertTrue(pe.getMessage().contains("cannot be an empty"));
-        }
-        Assert.assertTrue(thrown);
-        thrown = false;
-
-        //MISSING LENGTH PARAMETER
-        try {
-            run("FILL_RIGHT('123',foo)", variableMap);
-        }catch(ParseException pe){
-            thrown = true;
-            Assert.assertTrue(pe.getMessage().contains("expects three"));
-        }
-        Assert.assertTrue(thrown);
+    Assert.assertTrue(thrown);
+    thrown = false;
+
+    // NULL FILL
+    try{
+      run("FILL_RIGHT('123',foo, 7)", variableMap);
+    }catch(ParseException pe) {
+      thrown = true;
+      Assert.assertTrue(pe.getMessage().contains("are both required"));
+    }
+    Assert.assertTrue(thrown);
+    thrown = false;
+
+    // NON INTEGER LENGTH
+    try {
+      run("FILL_RIGHT('123','X', 'z' )", new HashedMap());
+    }catch(ParseException pe){
+      thrown = true;
+      Assert.assertTrue(pe.getMessage().contains("not a valid Integer"));
+    }
+    Assert.assertTrue(thrown);
+    thrown = false;
+
+    // EMPTY STRING PAD
+    try {
+      Object returnValue = run("FILL_RIGHT('123','', 10 )", new HashedMap());
+    }catch(ParseException pe) {
+      thrown = true;
+      Assert.assertTrue(pe.getMessage().contains("cannot be an empty"));
+    }
+    Assert.assertTrue(thrown);
+    thrown = false;
+
+    //MISSING LENGTH PARAMETER
+    try {
+      run("FILL_RIGHT('123',foo)", variableMap);
+    }catch(ParseException pe){
+      thrown = true;
+      Assert.assertTrue(pe.getMessage().contains("expects three"));
     }
+    Assert.assertTrue(thrown);
+  }
+
+  @Test
+  public void shannonEntropyTest() throws Exception {
+    //test empty string
+    Assert.assertEquals(0.0, (Double)run("STRING_ENTROPY('')", new HashMap<>()), 0.0);
+    Assert.assertEquals(0.0, (Double)run("STRING_ENTROPY(foo)", ImmutableMap.of("foo", "")), 0.0);
+
+    /*
+    Now consider the string aaaaaaaaaabbbbbccccc or 10 a's followed by 5 b's and 5 c's.
+    The probabilities of each character is as follows:
+    p(a) = 1/2
+    p(b) = 1/4
+    p(c) = 1/4
+    so the shannon entropy should be
+      -p(a)*log_2(p(a)) - p(b)*log_2(p(b)) - p(c)*log_2(p(c)) =
+      -0.5*-1 - 0.25*-2 - 0.25*-2 = 1.5
+     */
+    Assert.assertEquals(1.5, (Double)run("STRING_ENTROPY(foo)", ImmutableMap.of("foo", "aaaaaaaaaabbbbbccccc")), 0.0);
```

Comment:
It does seem that an empty string should have entropy exactly == 0.0 (line 160).

In line 173, I assume the argument is that a string with per-letter probabilities that are small exact powers of 2 will have its entropy calculated exactly.  I suspect that's true, but one would have to know the implementation to be sure.

Otherwise one would have to use a delta of 4*Math.ulp(Double(1.5)) , I think.
But it must pass as is, so it's good :-)